backend/src/apiserver/server/util.go

```diff
@@ -216,9 +216,25 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
+func validatePipelineSpecAndCheckPipelineVersionReference(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	pipelineVersionId, err := CheckPipelineVersionReference(resourceManager, resourceReferences)
+	if err != nil {
+		return err
+	}
+	if (spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "")) && pipelineVersionId == "" {
+		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
+	}
+	if spec != nil && spec.GetWorkflowManifest() != "" && pipelineVersionId != "" {
+		return util.NewInvalidInputError("Please either specify a pipeline version or a workflow manifest, not both.")
+	}
+	if err := ValidatePipelineSpec(resourceManager, spec); err != nil {
+		return err
+	}
+	return nil
```

Comment:
nit: the if conditions seem a little hard to parse, what do you think about the following?
```suggestion
	// centralize all the nil checks here
	pipelineId := spec != nil ? spec.GetPipelineId() : ""
	workflowManifest := spec != nil ? spec.GetWorkflowManifest() : ""
	pipelineVersionId, err := CheckPipelineVersionReference(resourceManager, resourceReferences)
	if err != nil {
		return err
	}
	if pipelineVersionId != "" || pipelineId != "" {
	    if workflowManifest != "" {
	        return util.NewInvalidInputError("Please either specify a pipeline version or a workflow manifest, not both.")
	    }
	    if pipelineVersionId != "" {
	    	// TODO: shall we verify pipelineId should be parent of this version here?
	    	return nil
	    } else {
	        return ValidatePipelineId(pipelineId)
	    }
	} else if workflowManifest != "" {
	    return ValidateWorkflowManifest(workflowManifest)
	} else {
	    // workflowManifest == "" && pipelineId == "" && pipelineVersionId == ""
	    return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
	}
```
the main ideas:
* give a name to partial results, so conditions are simpler.
* add some comments to clarify assumptions at different stages of the code.
* use nested if, so that it's clear we didn't forget any situation
* I think it makes more sense to break `ValidatePipelineSpec` into individual functions that either validate workflow manifests, or validates pipeline id.

also, there are similar if else checks in `ValidatePipelineSpec`, we no longer need them when centralizing all the conditions here.

Reply:
I think it may be more clear to differentiate the case by workflowmanifest is euqal to "" or not